lib/dao.js

```diff
@@ -1748,7 +1763,14 @@ DataAccessObject.prototype.save = function (options, cb) {
 
     inst.isValid(function (valid) {
       if (valid) {
-        save();
+        Model.notifyObserversOf('persist', context, function(err) {
+          if (err) return cb(err);
+
+          data = inst.toObject(true);
+          Model.applyProperties(data, inst);
+          inst.setAttributes(data);
```

Comment:
Without studying the source code extensively, I would expect that save performs the following steps (see also https://github.com/strongloop/loopback-datasource-juggler/pull/599#issuecomment-105183289):

```
 1. call "before save" hook with "ctx.instance"
 2. validate the instance
 3. convert the instance to a data object to be passed to the connector 
    - inst.toObject() + removeUndefined()
 4. call "persist" hook with the data from step 3
 5. call connector method with the data from step 3
 --- (async invocation of callback arg "saveCallback")--
 6. call "loaded" hook with data returned by the connector
 7. call inst._initProperties(data, { persisted: true });
 8. cal "after save" hook with "ctx.instance"
```
 
I think the reason why you need to add these extra three lines of code is that you are passing the instance to the hook, whereas my vision is to pass the raw data instead.

Here are few reasons why I think it's better to pass the data:
 - Changes made by the "persist" hook don't affect the original instance. If the DAO method does not update the instance with the data returned by the database, then the persist hook modifies only database representation, which is IMO exactly the intent of this hook.
 - My belief is that a model instance is always valid in an ideal world. The hook "persist" may modify some property values in such way that they no longer pass validation (for example hashed passwords may not meet rules like "must contain a lowercase and an uppercase character"). In that case the hook should not modify (and invalidate) the instance, but merely alter the data passed to the connector.